### PR TITLE
Grand circuit pathplanning test

### PR DESCRIPTION
**Pull request by GearBoxFox**

### CODE REVIEW

File: src/main/java/frc/robot/Constants.java [hunk 1]
```diff
@@ -146,7 +146,7 @@ public static class ArmConstants {
 
         public static final double ARM_OFFSET = 84 + 120 + 243;
 
-        public static final int ENCODER_PORT = 1;
+        public static final int ENCODER_PORT = 24;
```
Comment (Zoarial94) [inline src/main/java/frc/robot/Constants.java]: What encoder more specifically?

**Review by Zoarial94 — CHANGES_REQUESTED** [review]

**Comment by Zoarial94** [comment]
@GearBoxFox instead of having manual position control while balancing, can you have the path planner keep the current position?

File: src/main/java/frc/robot/commands/AutoBalance.java [hunk 2]
```diff
@@ -52,35 +61,53 @@ public void execute() {
 
     m_drivePower = Math.min(m_balanceController.calculate(m_currentAngle), 1);
     // Limit max power
-    m_drivePower = MathUtil.clamp(m_drivePower, -0.6, 0.6);
+    m_drivePower = MathUtil.clamp(m_drivePower, -0.5, 0.5);
```
Comment (Zoarial94) [inline src/main/java/frc/robot/commands/AutoBalance.java]: Math.min isn't doing anything here.

File: src/main/java/frc/robot/commands/AutoBalance.java [hunk 3]
```diff
@@ -52,35 +61,53 @@ public void execute() {
...
       // If robot has been balanced for 10 seconds, stop the command
       m_timer.stop();
       m_isLevel = true;
-    } else if (m_timer.get() > 0 && Math.abs(m_currentAngle) > AutoConstants.DESIRED_BALANCE_ANGLE) {
+      SmartDashboard.putBoolean("Timer Started", false);
+    }
+    if (Math.abs(m_error) < AutoConstants.ACCEPTABLE_BALANCE_ANGLE && m_timer.get() < 2) {
+      m_drivePower = 0.0;
+    }
+    if (m_timer.hasElapsed(0) && Math.abs(m_error) > AutoConstants.ACCEPTABLE_BALANCE_ANGLE) {
```
Comment (Zoarial94) [inline src/main/java/frc/robot/commands/AutoBalance.java:93]: These if statements should be mutually exclusive. You don't want unknowingly take multiple paths. You may need to reorder the statements as well to get the correct behavior.

File: src/main/java/frc/robot/subsystems/swerve/SwerveDrivetrain.java [hunk 4]
```diff
@@ -204,8 +204,12 @@ public Rotation2d getGyroPitch() {
...
     }
 
-    public double getGyroRollPitch() {
-        return m_currentRoll - m_previousRoll;
+    public Rotation2d getGyroRoll() {
+        return Rotation2d.fromDegrees(m_gyro.getRoll());
+    }
+
+    public double getGyroPitchRate() {
+        return m_currentPitch - m_previousPitch;
```
Comment (Zoarial94) [inline src/main/java/frc/robot/subsystems/swerve/SwerveDrivetrain.java]: This isn't really the pitch rate unless it accounts for time.

**Review by Zoarial94 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/commands/AutoBalance.java [hunk 5]
```diff
@@ -52,35 +61,53 @@ public void execute() {
...
+    }
+
+    if (m_balanceTimer.hasElapsed(0.01)) {
+//      m_drivePower = Math.copySign(0.15, -m_drive.getGyroPitch().getDegrees());
       m_drivePower = 0.0;
-//      SmartDashboard.putNumber("Stop Balancing", 1);
+    }
+    if (m_balanceTimer.hasElapsed(0.75)) {
+      m_balanceTimer.stop();
+      m_balanceTimer.reset();
```
Comment (Zoarial94) [inline src/main/java/frc/robot/commands/AutoBalance.java:74]: This shouldn't be timer based, it should be based off the robot pitch movement.

File: src/main/java/frc/robot/commands/AutoBalance.java (hunk 2 above)
Comment (GearBoxFox) [reply src/main/java/frc/robot/commands/AutoBalance.java]: This is the older version, this isn't being used anymore. The newer balance command is AutoBalanceTransCommand.

File: src/main/java/frc/robot/commands/AutoBalance.java (hunk 3 above)
Comment (GearBoxFox) [reply src/main/java/frc/robot/commands/AutoBalance.java:93]: Ok, I think we should merge the PR before we resolve this, as like i said earlier this version of the command isn't being used anymore.

File: src/main/java/frc/robot/commands/AutoBalance.java (hunk 5 above)
Comment (GearBoxFox) [reply src/main/java/frc/robot/commands/AutoBalance.java:74]: That can be done.

File: src/main/java/frc/robot/subsystems/swerve/SwerveDrivetrain.java (hunk 4 above)
Comment (GearBoxFox) [reply src/main/java/frc/robot/subsystems/swerve/SwerveDrivetrain.java]: I can update the method to take time into account

File: src/main/java/frc/robot/subsystems/swerve/SwerveDrivetrain.java [hunk 6]
```diff
@@ -196,16 +210,25 @@ public void setModuleStates(SwerveModuleState[] states) {
...
     }
 
-    public double getGyroRollPitch() {
-        return m_currentRoll - m_previousRoll;
+    public Rotation2d getGyroRoll() {
+        return Rotation2d.fromDegrees(m_gyro.getRoll());
+    }
+
+    public double getGyroPitchRate() {
+        double fpgaElapsedTime = RobotController.getFPGATime() - m_prevTime;
```
Comment (Zoarial94) [inline src/main/java/frc/robot/subsystems/swerve/SwerveDrivetrain.java:235]: We need to handle the edge case where m_prevTime == 0. If it is zero, set m_prevTime to the current time.

File: src/main/java/frc/robot/subsystems/swerve/SwerveDrivetrain.java [hunk 7]
```diff
@@ -196,16 +210,25 @@ public void setModuleStates(SwerveModuleState[] states) {
...
+        return (m_currentPitch - m_previousPitch) / RobotController.getFPGATime();
```
Comment (Zoarial94) [inline src/main/java/frc/robot/subsystems/swerve/SwerveDrivetrain.java]: Divide by the difference.

File: src/main/java/frc/robot/commands/AutoBalance.java [hunk 8]
```diff
@@ -41,46 +38,71 @@ public AutoBalance(SwerveDrivetrain drive) {
...
+    double m_error = AutoConstants.DESIRED_BALANCE_ANGLE - m_currentAngle;
+    double m_drivePower = Math.min(m_balanceController.calculate(m_currentAngle), 1);
 
-    m_drivePower = Math.min(m_balanceController.calculate(m_currentAngle), 1);
     // Limit max power
-    m_drivePower = MathUtil.clamp(m_drivePower, -0.6, 0.6);
+    m_drivePower = MathUtil.clamp(m_drivePower, -0.5, 0.5);
 
-    double rollRateBalance = m_drive.getGyroRollPitch();
+    double rollRateBalance = m_drive.getGyroPitchRate();
```
Comment (Zoarial94) [inline src/main/java/frc/robot/commands/AutoBalance.java:61]: The number returned here is different now that it is adjusted for time elapsed.

Comment (GearBoxFox) [reply src/main/java/frc/robot/commands/AutoBalance.java:61]: wouldn't it still be the rate of the roll, just properly calculated now as we're accouting for time elapsed?

Comment (Zoarial94) [reply src/main/java/frc/robot/commands/AutoBalance.java:61]: Correct. Just making a note for later.

**Comment by GearBoxFox** [comment]
@Zoarial94 I finished those last two things you requested, am I good to merge into main?

**Review by Zoarial94 — APPROVED** [review]
Looks good. We need to test ASAP though.

**Comment by GearBoxFox** [comment]
Yes, we can run another system check when we get to the pit.

**Pull request merged**